shared/src/main/java/com/couchbase/lite/PropertiesImpl.java

```diff
@@ -58,26 +64,65 @@ public Object get(String key) {
 
     @Override
     public String getString(String key) {
-        if (properties != null)
-            return (String) properties.get(key);
-        else
-            return root == null ? null : root.get(key).asString();
+        if (properties != null) {
+            Object obj = properties.get(key);
+            if (obj != null && obj instanceof String)
+                return (String) obj;
+            else
+                return null;
+        } else {
+            if (root == null)
+                return null;
+            else {
+                FLValue flvalue = root.get(key);
+                if (flvalue.getType() == kFLString)
+                    return flvalue.asString();
+                else
+                    return null;
+            }
+        }
     }
 
     @Override
     public int getInt(String key) {
-        if (properties != null)
-            return (int) properties.get(key);
-        else
-            return root == null ? 0 : root.get(key).asInt();
+        if (properties != null) {
+            Object obj = properties.get(key);
+            if (obj != null && obj instanceof Number)
+                return ((Number) obj).intValue();
+            else
+                return 0;
+        } else {
+            if (root == null)
+                return 0;
+            else {
+                FLValue flvalue = root.get(key);
+                if (flvalue.getType() == kFLNumber)
+                    return flvalue.asInt();
+                else
+                    return 0;
+            }
+        }
     }
```

Comment:
Regarding the second comment, current Obj-C implementation does not check the FLValue type. It might return unexpected value

```Obj-C
- (double) doubleForKey: (NSString*)key {
    if (_properties)
        return [numberProperty(_properties, key) doubleValue];
    else
        return FLValue_AsDouble([self fleeceValueForKey: key]);
}

- (float) floatForKey: (NSString*)key {
    if (_properties)
        return [numberProperty(_properties, key) floatValue];
    else
        return FLValue_AsFloat([self fleeceValueForKey: key]);
}

- (NSInteger) integerForKey: (NSString*)key {
    if (_properties)
        return [numberProperty(_properties, key) integerValue];
    else
        return (NSInteger)FLValue_AsInt([self fleeceValueForKey: key]);
}
```

```C
bool FLValue_AsBool(FLValue v)                  {return v && v->asBool();}
int64_t FLValue_AsInt(FLValue v)                {return v ? v->asInt() : 0;}
uint64_t FLValue_AsUnsigned(FLValue v)          {return v ? v->asUnsigned() : 0;}
float FLValue_AsFloat(FLValue v)                {return v ? v->asFloat() : 0.0;}
double FLValue_AsDouble(FLValue v)              {return v ? v->asDouble() : 0.0;}
FLSlice FLValue_AsString(FLValue v)             {return v ? external(v->asString()) : kFLSliceNull;}
FLSlice FLValue_AsData(FLValue v)               {return v ? external(v->asData()) : kFLSliceNull;}
FLArray FLValue_AsArray(FLValue v)              {return v ? v->asArray() : nullptr;}
FLDict FLValue_AsDict(FLValue v)                {return v ? v->asDict() : nullptr;}
```